src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/SpecialFileProviders/AbstractSpecialFileProvider.cs

```
@@ -11,37 +15,97 @@ namespace Microsoft.VisualStudio.ProjectSystem.SpecialFileProviders
-                {
-                    await CreateFileAsync(path);
-                }
+        private async Task<string?> CreateDefaultFileAsync(IProjectTreeProvider provider, IProjectTree root, SpecialFileFlags flags)
+        {
+            string? path = await GetDefaultFileAsync(provider, root);
+
+            if (path != null && flags.HasFlag(SpecialFileFlags.CreateIfNotExist))
+            {
+                await CreateFileAsync(path);
             }
 
             // We always return the default path, regardless of whether we created it or it exists, as per contract
             return path;
         }
 
-        protected abstract Task<string?> FindFileAsync(IProjectTreeProvider provider, IProjectTree root);
+        protected abstract Task<IProjectTree?> FindFileAsync(IProjectTreeProvider provider, IProjectTree root);
+
+        protected string? GetFilePath(IProjectTreeProvider provider, IProjectTree node)
+        {
+            return _isFolder ? provider.GetRootedAddNewItemDirectory(node) : provider.GetPath(node);
+        }
 
         protected abstract Task<string?> GetDefaultFileAsync(IProjectTreeProvider provider, IProjectTree root);
 
-        protected abstract Task CreateFileAsync(string path);
+        protected virtual Task CreateFileAsync(string path)
+        {
+            return _isFolder ? _storage.CreateFolderAsync(path) : _storage.CreateEmptyFileAsync(path);
+        }
+
+        private Task AddFileAsync(string path)
+        {
+            return _isFolder ? _storage.AddFolderAsync(path) : _storage.AddFileAsync(path);
+        }
+
+        private Task VerifyStateAsync(IProjectTree node, string path)
+        {
+            if (_isFolder != node.IsFolder)
+                throw new IOException(string.Format(CultureInfo.CurrentCulture, Resources.SpecialFileProvider_FileOrFolderAlreadyExists, node.Caption), Win32Interop.HResultFromWin32(Win32Interop.ERROR_FILE_EXISTS));
```

This is supposed to be shown to the user, but the CPS is currently hiding errors from special file handlers: http://index/?query=ProjectNode.GetFile&rightProject=Microsoft.VisualStudio.ProjectSystem.VS.Implementation&file=Package%5CNodes%5CProjectNode.cs&line=2651.